common/switchboard.hpp

```diff
@@ -1,179 +1,571 @@
+ *     topic2.put(event2);
  * }
+ * 
+ * // Read topic 3 synchronously
+ * switchboard.schedule<topic3_type>(plugin_id, "topic3", [&](switchboard::ptr<topic3_type> event3, std::size_t it) {
+ *     // This is a lambda expression
+ *     // https://en.cppreference.com/w/cpp/language/lambda
+ *     std::cout << "Got a new event on topic3: " << event3->foo << " for iteration " << it << std::endl;
+ * });
  * \endcode
- *
  */
 class switchboard : public phonebook::service {
+public:
+
+    /**
+     * @brief The type of shared pointer returned by switchboard.
+     *
+     * TODO: Make this agnostic to the type of `ptr`
+     * Currently, it depends on `ptr` == shared_ptr
+     */
+    template <typename specific_event>
+    using ptr = std::shared_ptr<specific_event>;
+
+    class event {
+    public:
+        virtual ~event() { }
+    };
+
+    /**
+     * @brief Helper class for making event types
+     *
+     * Since topic has no static type-information on the contained events, this class does not
+     * either.
+     *
+     * \code{.cpp}
+     * event_wrapper<int> int_event = 5;
+     * \endcode
+     */
+    template <typename underlying_type>
+    class event_wrapper : public event {
+    private:
+        underlying_type underlying_data;
+    public:
+        event_wrapper() { }
+
+        event_wrapper(underlying_type underlying_data_)
+            : underlying_data{underlying_data_}
+        { }
+        operator underlying_type() const { return underlying_data; }
+        underlying_type& operator*() { return underlying_data; }
+        const underlying_type& operator*() const { return underlying_data; }
+    };
 
 private:
-	virtual
-	std::unique_ptr<writer<void>> _p_publish(const std::string& topic_name, std::size_t ty) = 0;
+    /**
+     * @brief Represents a single topic_subscription (callback and queue)
+     *
+     * This class treats everything as `event`s (type-erased) because `topic` treats everything as
+     * `event`s.
+     *
+     * Each topic can have 0 or more topic_subscriptions.
+     */
+    class topic_subscription {
+    private:
+        const std::string& _m_topic_name;
+        plugin_id_t _m_plugin_id;
+        std::function<void(ptr<const event>&&, std::size_t)> _m_callback;
+        const std::shared_ptr<record_logger> _m_record_logger;
+        record_coalescer _m_cb_log;
+        moodycamel::BlockingConcurrentQueue<ptr<const event>> _m_queue {8 /*max size estimate*/};
+        moodycamel::ConsumerToken _m_ctok {_m_queue};
+        static constexpr std::chrono::milliseconds _m_queue_timeout {100};
+        std::size_t _m_enqueued {0};
+        std::size_t _m_dequeued {0};
+        std::size_t _m_idle_cycles {0};
+
+        // This needs to be last,
+        // so it is destructed before the data it uses.
+        managed_thread _m_thread;
+
+        void thread_on_start() {
+#ifndef NDEBUG
+            // std::cerr << "Thread " << std::this_thread::get_id() << " start" << std::endl;
```

Comment:
I think @charmoniumQ wants to avoid over-logging. I'm generally okay with verbose output under `dbg`. Maybe we can leave it commented until we have flags for verbose logging? Currently, the `ENABLE_VERBOSE_ERRORS` flag is only for errors (`common/error_util.hpp`).

Reply:
We can uncomment that since it does not print in steady-state.